### ISSUE TITLE
Should casts establish context for downward inference?

### ISSUE THREAD

**Issue by vsmenon** [dart-lang/sdk#30235]
The following code (from an app) throws an exception: "CastError: Casting value of type '_FrozenElementList<Element>' to incompatible type 'List<HtmlElement>'" 

```dart
 List<HtmlElement> _getRowCells(Element row) {
    return row.querySelectorAll("div") as List<HtmlElement>;
  }
```

Note, that `querySelectorAll` is generic and if the cast is simply omitted, its type is inferred as `HtmlElement`.

Should the cast also establish downward context?

@leafpetersen

**Comment by jmesserly** [comment]
indeed, we previously did this, and then it was removed: https://github.com/dart-lang/sdk/commit/7ccd8dbfe45236648532221bc00c57ed71a3ded1, https://github.com/dart-lang/sdk/issues/26027 for previous discussion

**Comment by jmesserly** [comment]
The reason we changed this is:
```dart
var cells = row.querySelectorAll("div") as List<HtmlElement>;
```
if we infer from the `as`, then the runtime cast ends up unnecessary. Ideally we could skip the message in that case, but it's a bit hard to know if the `as` is what made inference succeed or not.

**Comment by leafpetersen** [comment]
I'm still a bit on the fence with this one.  Ignoring the pragmatics for a moment (i.e. that it's a bit difficult to give unnecessary cast warnings if we infer from casts):

On the plus side, the cast does provide an expectation, so there's no technical reason not to propagate it

On the minus side, there are very few situations in which this is the right thing to do.  You should be always able to either use a type on a variable or explicitly pass type arguments.  The only exceptions would be places where there is no syntax for writing down a type (i.e. function return types, and for the time being at least partial generic instantiations).

**Comment by jmesserly** [comment]
Yeah agreed, also on the fence.

I wonder how often folks do this because of the learning curve for generic methods (including the API changes e.g. querySelectorAll). Runtime casts shouldn't be needed so much anymore in strong mode, and they're more likely to fail, so it's kinda nice to discourage `as` when it's not necessary.

On the other hand, we could certainly propagate that type, and like our other downwards inference, it works well at reducing the odds of runtime cast failures, and improving the odds of static error messages. Generally that's been a good thing & a big help in migrating code to strong mode.

FWIW, I lean slightly towards using the `as` context type, and see if we can fix the "unnecessary cast" messages.

**Comment by lrhn** [comment]
I'm all for something like this, but it didn't make it into Dart 2.
It's a breaking change if we change the behavior in a future version, unless we can ensure that it only changes behavior in cases that would otherwise have thrown at runtime anyway.
